src/ecdsa/recovery.rs

```diff
@@ -253,8 +254,9 @@ mod tests {
         assert!(vrfy.recover_ecdsa(&msg, &sigr).is_ok());
         assert!(full.recover_ecdsa(&msg, &sigr).is_ok());
 
-        assert_eq!(vrfy.recover_ecdsa(&msg, &sigr), full.recover_ecdsa(&msg, &sigr));
-        assert_eq!(full.recover_ecdsa(&msg, &sigr), Ok(pk));
+        let vrfy_res = vrfy.recover_ecdsa(&msg, &sigr);
+        let full_res = full.recover_ecdsa(&msg, &sigr);
+        assert!(matches!(vrfy_res, full_res));
```

Comment:
In 645238fd761afe2e95925a0108cef937e243012b:

This is not correct -- you can't use `matches!` on two variables like this; the result is to bind `vrfy_res` to `full_res` and then check that the binding succeeded (which it always will because there are no further restrictions). See https://stackoverflow.com/questions/72537643/apparent-unused-variable-in-match-statement

In general I'm a bit suspicious of these `allow(unused_variables)` lines because I think they're masking similar mistakes .... and in retrospect we should revisit what's going on in rust-bitcoin with these.

Reply:
ok, I need to read the `matches` code then before using it willy-nilly. Just had a thought, perhaps we could do this on error types

```rust
#[derive(Debug)]
#[non_exhaustive]
#[allow(missing_copy_implementations)]
#[cfg_attr(test, derive(Copy, PartialEq, Eq, Clone))]
```

Then all the test changes in this PR are not required and `matches!` use goes away.

Reply:
Why are we removing `PartialEq` **in `secp256k1`**? The biggest reason to remove it is that we may add `std::io::Error` inside one day but I don't expect this to happen in *this* crate. Same with `Clone`. I'd avoid `Copy` since adding something that allocates is quite plausible.

Reply:
Ah, that's a good point -- I had half-forgotten that the reason for "error types don't impl any traits" is that `io::Error` tends to poison our types.  But you're right that we're extremely unlikely to ever have I/O errors from this pure-math crate.

Reply:
Oh, I thought that the "great error cleanup", across the org, meant that all errors should only derive `Debug`? Are you guys saying that for every error type I/we have to think about what traits to derive? That sounds like unnecessary work, can't we just decide on a minimum set and use that (which is what I thought we did and how we got to only deriving `Debug`)?

Reply:
Heh, this is also a good pointt.... I certainly don't think we should have to think hard about every error type. That's not sustainable.

But OTOH maybe for "edge crates" like rust-secp and rust-bech32, which do pure computation, we could have a more generous set of error #derives .... I dunno :/.

Reply:
As a thought exercise, what benefit does having the extra derives give? My thoughts on this this morning were "as a user, if I'm used to only ever having `Debug` i.e., no `Eq` when working with `rust-bitcoin` crates I'd probably not bother checking the specific error" - if this is true, then it doesn't add much benefit and reduces maintainability (admittedly by a very small amount). Thoughts?

Reply:
Being unable to compare errors is *awful*. It's awful in unit tests and it's awful when checking error returns. People only live with it because `std` forces it on people and `std` has a many-year history of ignoring user feedback.

Reply:
> for every error type I/we have to think about what traits to derive?

Yes. The rule is basically "don't make promises you can't keep".



> That sounds like unnecessary work, can't we just decide on a minimum set and use that

Unfortunately, that's the reality of programming. Or maybe you could argue the consequence of widespread, terrible `io::Error`. Either way, there's no reasonable solution. If you don't want to think about it I'll be happy to do this thinking instead.

> Being unable to compare errors is _awful_. It's awful in unit tests and it's awful when checking error returns. People only live with it because `std` forces it on people

Exactly

> `std` has a many-year history of ignoring user feedback.

I don't think so. It's just that this part of `std` got stabilized (maybe too quickly?) and painted itself into a corner. (Same with a few other parts.)

Anyway, if my proposal of having non-std IO trait goes through we won't have any IO error and instead have the errors generic which would imply we can actually derive `Clone` and `PartialEq` for all error types.

Reply:
>  we can actually derive Clone and PartialEq for all error types.

I'd be super happy with this.

Reply:
Curios why `Eq` is excluded from the list

Reply:
Just lazy typing by me, I should have written `PartialEq`/`Eq` :)

Reply:
oh I didn't type it, lazy typing by @Kixunil (I assume) - double smiley face.

Reply:
Not necessarily. We may avoid promising `Eq` to avoid future breakage like what happened with `LockTime`. But I'm not entirely sure about this one, there's a good chance `Eq` is just fine.

Reply:
The `LockTime` breakage was about `Ord`. Having a type that is `PartialEq` but not `Eq` is a pretty obscure use-case, and I'm doubtful that crypto errors will meet it.

Reply:
Indeed, I'm very open to having `Eq`.